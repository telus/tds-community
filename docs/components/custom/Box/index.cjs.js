require('./dist/index.css')
const Box = require('./dist/index.cjs')

module.exports = Box
