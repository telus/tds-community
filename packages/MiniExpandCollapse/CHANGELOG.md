# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2020-09-09)


### Bug Fixes

* **community-mini-expand-collapse:** update prop types ([f864436](https://github.com/telus/tds-community/commit/f864436b4d151692e865c16bba2d28b11d51cda7))
* **community-mini-expand-collapse:** update proptype ([622fd7f](https://github.com/telus/tds-community/commit/622fd7ff6cdc61e7a8538cdd5d056974ce93042d))


### Features

* **community-mini-expand-collapse:** mini expand collapse ([deb3a07](https://github.com/telus/tds-community/commit/deb3a07f78ea4d56cdcea49a41eb09da90605991))





# 0.1.0 (2020-09-09)


### Bug Fixes

* **community-mini-expand-collapse:** update prop types ([f864436](https://github.com/telus/tds-community/commit/f864436b4d151692e865c16bba2d28b11d51cda7))
* **community-mini-expand-collapse:** update proptype ([622fd7f](https://github.com/telus/tds-community/commit/622fd7ff6cdc61e7a8538cdd5d056974ce93042d))


### Features

* **community-mini-expand-collapse:** mini expand collapse ([deb3a07](https://github.com/telus/tds-community/commit/deb3a07f78ea4d56cdcea49a41eb09da90605991))
