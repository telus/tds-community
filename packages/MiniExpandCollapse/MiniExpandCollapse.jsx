import React, { useRef, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { FadeAndReveal } from '@tds/shared-animation'
import { responsiveProps } from '@tds/util-prop-types'

import Link from './Link/Link'

/**
 * A hyperlink to expand target content currently not in view.
 * @version ./package.json
 * @visibleName MiniExpandCollapse (beta)
 */
const MiniExpandCollapse = ({ children, expandTitle, collapseTitle, size, invert, a11yLabel }) => {
  const contentRef = useRef(null)
  const [isPanelOpen, setPanel] = useState(false)
  const [contentHeight, setContentHeight] = useState(0)
  const togglePanel = () => {
    setPanel(!isPanelOpen)
  }
  useEffect(() => {
    if (contentRef.current.offsetHeight !== contentHeight) {
      setContentHeight(contentRef.current.offsetHeight)
    }
  }, [contentHeight])

  return (
    <div>
      <Link
        expandTitle={expandTitle}
        collapseTitle={collapseTitle}
        isPanelOpen={isPanelOpen}
        togglePanel={togglePanel}
        size={size}
        invert={invert}
        a11yLabel={a11yLabel}
      />
      <FadeAndReveal timeout={500} in={isPanelOpen} height={contentHeight} duration={500}>
        {() => <div ref={contentRef}>{children}</div>}
      </FadeAndReveal>
    </div>
  )
}

MiniExpandCollapse.propTypes = {
  /**
   * Link text when content is collapsed/hidden.
   */
  collapseTitle: PropTypes.string.isRequired,
  /**
   * Link text when target content is expanded. Defaults to collapse title.
   */
  expandTitle: PropTypes.string,
  /**
   * Required. Size of link text and icon. One of `small, medium, large` as a [**responsive prop**](#responsiveProps).
   */
  size: responsiveProps(PropTypes.oneOf(['small', 'medium', 'large'])).isRequired,
  /**
   * Invert the style to appear light on dark backgrounds.
   */
  invert: PropTypes.bool,
  /**
   * Additional title text to provide more context for users using screen readers.
   */
  a11yLabel: PropTypes.string,
  /**
   * Target content to expand downward. Can be text, any HTML element, or any component.
   */
  children: PropTypes.node.isRequired,
}

MiniExpandCollapse.defaultProps = {
  expandTitle: null,
  invert: false,
  a11yLabel: undefined,
}

export default MiniExpandCollapse
