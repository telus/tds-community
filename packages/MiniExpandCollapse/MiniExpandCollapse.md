Similar to ExpandCollapse. Clicking on the hyperlink triggers target content to display in a panel directly below CTA.

### Usage criteria

- Mainly used to answer a question potentially posed by a user regarding information they could interpret as unclear
- Should provide more details than a Tooltip
- Expand/collapse titles should be kept short and if possible, sentence case should be used
- Titles should be lead by a strong action word
- Note: unlike `ExpandCollapse`, the panel housing the target content has no inherent spacing

```jsx
<MiniExpandCollapse
  collapseTitle="Why do I see a $5 one-time credit?"
  size={{ xs: 'small', md: 'medium' }}
>
  <Box vertical={4}>
    <Paragraph size="medium">
      As a TELUS customer, you pay one month in advance for your upcoming services. Because you
      changed your services mid-way through your billing cycle, we will:
    </Paragraph>
    <br />
    <UnorderedList size="medium">
      <UnorderedList.Item>
        Credit you for XX days you already paid for with you previous plan, but didn't use
      </UnorderedList.Item>
      <UnorderedList.Item>
        Charge you for the remaining XX days of your billing cycle based on the new plan and add-ons
        you just activated.
      </UnorderedList.Item>
    </UnorderedList>
    <br />
    <Paragraph size="medium">
      The difference between these amount is what forms the one-time credit you'll see on your next
      bill only.
    </Paragraph>
  </Box>
</MiniExpandCollapse>
```

```jsx
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col xs={12} md={4}>
      <MiniExpandCollapse
        expandTitle="Hide details"
        collapseTitle="Show details"
        size={{ xs: 'medium', md: 'small' }}
        a11yLabel="Details regarding charges"
      >
        <Box vertical={3}>
          <HairlineDivider />
          <Box between={2}>
            <Box between="space-between" inline>
              <Paragraph size="medium">Buries Drop Wire</Paragraph>
              <Paragraph size="medium">$2.00</Paragraph>
            </Box>
            <Box between="space-between" inline>
              <Paragraph size="medium">Jacks/Outlet</Paragraph>
              <Paragraph size="medium">$75.00</Paragraph>
            </Box>
            <Box between="space-between" inline>
              <Paragraph size="medium">UPS Battery</Paragraph>
              <Paragraph size="medium">$30.00</Paragraph>
            </Box>
          </Box>
        </Box>
      </MiniExpandCollapse>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

```jsx { "props": { "className": "docs_purple-block" } }
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col xs={12} md={4}>
      <MiniExpandCollapse
        expandTitle="Hide details"
        collapseTitle="Show details"
        size={{ xs: 'medium', md: 'small' }}
        invert
      >
        <Box vertical={3}>
          <HairlineDivider />
          <Box between={2}>
            <Box between="space-between" inline>
              <Paragraph size="medium" invert>
                Buries Drop Wire
              </Paragraph>
              <Paragraph size="medium" invert>
                $2.00
              </Paragraph>
            </Box>
            <Box between="space-between" inline>
              <Paragraph size="medium" invert>
                Jacks/Outlet
              </Paragraph>
              <Paragraph size="medium" invert>
                $75.00
              </Paragraph>
            </Box>
            <Box between="space-between" inline>
              <Paragraph size="medium" invert>
                UPS Battery
              </Paragraph>
              <Paragraph size="medium" invert>
                $30.00
              </Paragraph>
            </Box>
          </Box>
        </Box>
      </MiniExpandCollapse>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```
