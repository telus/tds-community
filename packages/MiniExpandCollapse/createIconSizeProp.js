const createIconSizeProp = size => {
  if (typeof size === 'object' && size !== null) {
    const iconSize = {}
    Object.keys(size).map(breakpoint => {
      if (size[breakpoint] === 'small') {
        iconSize[breakpoint] = 20
      } else {
        iconSize[breakpoint] = 24
      }
      return breakpoint
    })
    return iconSize
  }
  return size === 'small' ? 20 : 24
}

export default createIconSizeProp
