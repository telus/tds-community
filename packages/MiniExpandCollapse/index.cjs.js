const MiniExpandCollapse = require('./dist/index.cjs')

module.exports = MiniExpandCollapse
