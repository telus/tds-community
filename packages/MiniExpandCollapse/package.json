{
  "name": "@tds/community-mini-expand-collapse",
  "version": "0.1.0",
  "description": "",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telus/tds-community.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telus/tds-community/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^16.8.2",
    "react-dom": "^16.8.2",
    "styled-components": "^4.1.3"
  },
  "dependencies": {
    "@tds/core-a11y-content": "^2.1.5",
    "@tds/core-colours": "^2.2.1",
    "@tds/core-interactive-icon": "^2.1.0",
    "@tds/shared-animation": "^3.0.0",
    "@tds/shared-styles": "^1.5.2",
    "@tds/shared-typography": "^1.3.5",
    "@tds/util-helpers": "^1.5.0",
    "@tds/util-prop-types": "^1.4.0",
    "prop-types": "^15.6.2"
  }
}
