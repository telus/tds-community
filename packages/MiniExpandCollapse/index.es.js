import MiniExpandCollapse from './dist/index.es'

export default MiniExpandCollapse
