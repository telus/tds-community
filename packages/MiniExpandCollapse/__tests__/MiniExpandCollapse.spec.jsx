import React from 'react'
import { shallow, mount } from 'enzyme'
import Box from '@tds/core-box'
import Paragraph from '@tds/core-paragraph'

import MiniExpandCollapse from '../MiniExpandCollapse'

describe('MiniExpandCollapse', () => {
  const doShallow = () =>
    shallow(
      <MiniExpandCollapse expandTitle="Hide details" collapseTitle="Show details" size="large">
        <Box vertical={4}>
          <Paragraph>The content to be expanded.</Paragraph>
        </Box>
      </MiniExpandCollapse>
    )

  const doMount = () =>
    mount(
      <MiniExpandCollapse expandTitle="Hide details" collapseTitle="Show details" size="large">
        <Box vertical={4}>
          <Paragraph>The content to be expanded.</Paragraph>
        </Box>
      </MiniExpandCollapse>
    )

  it('renders correctly', () => {
    const miniExpandCollapse = doShallow()
    expect(miniExpandCollapse).toMatchSnapshot()
  })

  it('renders children', () => {
    const miniExpandCollapse = doMount()
    expect(miniExpandCollapse.find('Paragraph').exists()).toBeTruthy()
  })

  it('renders a title correctly', () => {
    const miniExpandCollapse = doMount()
    expect(
      miniExpandCollapse
        .find('Link')
        .children()
        .text()
    ).toEqual('Show details')
  })

  it('renders an interactive icon', () => {
    const miniExpandCollapse = doMount()
    expect(miniExpandCollapse.find('CaretDown').exists()).toBeTruthy()
  })

  it('renders A11y content', () => {
    const miniExpandCollapse = mount(
      <MiniExpandCollapse
        expandTitle="Hide details"
        collapseTitle="Show details"
        size="large"
        a11yLabel="detailed description"
      >
        <Box vertical={4}>
          <Paragraph>The content to be expanded.</Paragraph>
        </Box>
      </MiniExpandCollapse>
    )
    expect(miniExpandCollapse.find('A11yContent').exists()).toBeTruthy()
  })
})
