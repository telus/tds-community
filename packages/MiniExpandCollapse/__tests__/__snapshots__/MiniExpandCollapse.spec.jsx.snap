// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniExpandCollapse renders correctly 1`] = `
<div>
  <Link
    collapseTitle="Show details"
    expandTitle="Hide details"
    isPanelOpen={false}
    size="large"
    togglePanel={[Function]}
  />
  <FadeAndReveal
    duration={500}
    height={0}
    in={false}
    timeout={500}
  >
    <Component />
  </FadeAndReveal>
</div>
`;
